src/rpc/blockchain.cpp

```diff
@@ -2799,6 +2799,79 @@ static RPCHelpMan loadtxoutset()
+    {RPCResult::Type::NUM, "difficulty", "difficulty of the tip"},
+    {RPCResult::Type::NUM, "verificationprogress", "progress towards the network tip"},
+    {RPCResult::Type::STR_HEX, "snapshot_blockhash", /*optional=*/true, "the base block of the snapshot this chainstate is based on, if any"},
+    {RPCResult::Type::NUM, "coins_db_cache_bytes", "size of the coinsdb cache"},
+    {RPCResult::Type::NUM, "coins_tip_cache_bytes", "size of the coinstip cache"},
+};
+
+static RPCHelpMan getchainstates()
+{
+return RPCHelpMan{
+        "getchainstates",
+        "\nReturn information about chainstates.\n",
+        {},
+        RPCResult{
+            RPCResult::Type::OBJ, "", "", {
+                {RPCResult::Type::NUM, "headers", "the number of headers seen so far"},
+                {RPCResult::Type::OBJ, "normal", /*optional=*/true, "fully validated chainstate containing blocks this node has validated starting from the genesis block", RPCHelpForChainstate},
+                {RPCResult::Type::OBJ, "snapshot", /*optional=*/true, "only present if an assumeutxo snapshot is loaded. Partially validated chainstate containing blocks this node has validated starting from the snapshot. After the snapshot is validated (when the 'normal' chainstate advances far enough to validate it), this chainstate will replace and become the 'normal' chainstate.", RPCHelpForChainstate},
+            }
+        },
+        RPCExamples{
+            HelpExampleCli("getchainstates", "")
+    + HelpExampleRpc("getchainstates", "")
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    LOCK(cs_main);
+    UniValue obj(UniValue::VOBJ);
+
+    NodeContext& node = EnsureAnyNodeContext(request.context);
+    ChainstateManager& chainman = *node.chainman;
+
+    auto make_chain_data = [&](const Chainstate& cs) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+        AssertLockHeld(::cs_main);
+        UniValue data(UniValue::VOBJ);
+        if (!cs.m_chain.Tip()) {
+            return data;
+        }
+        const CChain& chain = cs.m_chain;
+        const CBlockIndex* tip = chain.Tip();
+
+        data.pushKV("blocks",                (int)chain.Height());
+        data.pushKV("bestblockhash",         tip->GetBlockHash().GetHex());
+        data.pushKV("difficulty",            (double)GetDifficulty(tip));
+        data.pushKV("verificationprogress",  GuessVerificationProgress(Params().TxData(), tip));
+        data.pushKV("coins_db_cache_bytes",  cs.m_coinsdb_cache_size_bytes);
+        data.pushKV("coins_tip_cache_bytes", cs.m_coinstip_cache_size_bytes);
+        if (cs.m_from_snapshot_blockhash) {
+            data.pushKV("snapshot_blockhash", cs.m_from_snapshot_blockhash->ToString());
+        }
+        return data;
+    };
+
+    if (chainman.GetAll().size() > 1) {
+        for (Chainstate* chainstate : chainman.GetAll()) {
+            obj.pushKV(
+                chainstate->m_from_snapshot_blockhash ? "snapshot" : "normal",
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1341995611

> It's kind of confusing that here in the results the ibd chainstate is also called normal. It might be good to add a comment on that.

@fjahr could you maybe suggest specific changes that you would recommend here? I'm confused by this comment and the one about removing the if block https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1342018556. Removing the if-block would not work (as far as I know) because it would cause the "snapshot" chainstate to stick around after it is validated, and the "normal" chainstate to disappear, instead of the "snapshot" chainstate replacing the "normal" chainstate when it is fully validated as intended.

The idea is for the "normal" chainstate to refer to a typical fully validated chainstate, and for the "snapshot" to refer to a temporarily assumed-valid chainstate loaded from a snapshot.

I think the term "IBD chainstate" is basically meaningless because both chainstates are downloaded during IBD. Even though the term is used internally in code, I think it would be good to move away from it and not expose it in the RPC interface.

Reply:
@ryanofsky Sure, first of all it sounds like we are on the same page that naming of the chainstates isn’t ideal. I think it will be harder for future developers to understand how assumeutxo works when snapshot and assumed_valid are used interchangably as well as ibd and background. In my mind is an area of improvement for the future but obviously the ship has sailed for this PR if we want to merge it any time soon and for James’ sanity also (and especially if this is merged while I write this :D ).

I think it makes sense to hide that complexity from the users but that rationale should be documented for developers in a comment here, and that’s what I was asking for. In the future I think we should try to make chainstate naming consistent and if we believe that calling ibd chainstates normal both in the context of assume UTXO as well as outside of it, then we may want to even think about a DDD approach and rename ibd/background chainstate to something like normal_background.

For the if-else: right, I didn’t take that transition period when we are fully synced but both chainstates are still around into account correctly. I guess it can still work with a guard clause but it’s not much of an improvement anymore then.